Failed to insert values into table store

```
org.apache.hadoop.service.AbstractService: Service org.apache.hadoop.mapreduce.v2.app.MRAppMaster failed in state INITED; cause: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.hive.ql.io.HiveFileFormatUtils$NullOutputCommitter not found
java.lang.RuntimeException: java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.hive.ql.io.HiveFileFormatUtils$NullOutputCommitter not found
	at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2229)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster$1.call(MRAppMaster.java:481)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster$1.call(MRAppMaster.java:458)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.callWithJobClassLoader(MRAppMaster.java:1560)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.createOutputCommitter(MRAppMaster.java:458)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.serviceInit(MRAppMaster.java:377)
	at org.apache.hadoop.service.AbstractService.init(AbstractService.java:163)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster$4.run(MRAppMaster.java:1518)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1727)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.initAndStartAppMaster(MRAppMaster.java:1515)
	at org.apache.hadoop.mapreduce.v2.app.MRAppMaster.main(MRAppMaster.java:1448)
Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: Class org.apache.hadoop.hive.ql.io.HiveFileFormatUtils$NullOutputCommitter not found
	at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2197)
	at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2221)
	... 12 more
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.hive.ql.io.HiveFileFormatUtils$NullOutputCommitter not found
	at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:2103)
	at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2195)
	... 13 more
```

--- a/emr-tablestore/src/main/java/com/aliyun/openservices/tablestore/hive/JarFinder.java
+++ b/emr-tablestore/src/main/java/com/aliyun/openservices/tablestore/hive/JarFinder.java
@@ -1,0 +1,217 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.aliyun.openservices.tablestore.hive;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLDecoder;
+import java.text.MessageFormat;
+import java.util.Enumeration;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import com.alicloud.openservices.tablestore.core.utils.Preconditions;
+
+/**
+ * Borrowed from https://github.com/apache/hbase/blob/master/hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/JarFinder.java
+ */
+public class JarFinder {
+  public JarFinder() {
+  }
+
+  private static void copyToZipStream(File file, ZipEntry entry, ZipOutputStream zos) throws IOException {
+    FileInputStream is = new FileInputStream(file);
+
+    try {
+      zos.putNextEntry(entry);
+      byte[] arr = new byte[4096];
+
+      for(int read = is.read(arr); read > -1; read = is.read(arr)) {
+        zos.write(arr, 0, read);
+      }
+    } finally {
+      try {
+        is.close();
+      } finally {
+        zos.closeEntry();
+      }
+    }
+
+  }
+
+  public static void jarDir(File dir, String relativePath, ZipOutputStream zos) throws IOException {
+    Preconditions.checkNotNull(relativePath, "relativePath");
+    Preconditions.checkNotNull(zos, "zos");
+    File manifestFile = new File(dir, "META-INF/MANIFEST.MF");
+    ZipEntry manifestEntry = new ZipEntry("META-INF/MANIFEST.MF");
+    if (!manifestFile.exists()) {
+      zos.putNextEntry(manifestEntry);
+      (new Manifest()).write(new BufferedOutputStream(zos));
+      zos.closeEntry();
+    } else {
+      copyToZipStream(manifestFile, manifestEntry, zos);
+    }
+
+    zos.closeEntry();
+    zipDir(dir, relativePath, zos, true);
+    zos.close();
+  }
+
+  private static void zipDir(File dir, String relativePath, ZipOutputStream zos, boolean start) throws IOException {
+    String[] dirList = dir.list();
+    if (dirList != null) {
+      String[] var5 = dirList;
+      int var6 = dirList.length;
+
+      for(int var7 = 0; var7 < var6; ++var7) {
+        String aDirList = var5[var7];
+        File f = new File(dir, aDirList);
+        if (!f.isHidden()) {
+          String filePath;
+          if (f.isDirectory()) {
+            if (!start) {
+              ZipEntry dirEntry = new ZipEntry(relativePath + f.getName() + "/");
+              zos.putNextEntry(dirEntry);
+              zos.closeEntry();
+            }
+
+            filePath = f.getPath();
+            File file = new File(filePath);
+            zipDir(file, relativePath + f.getName() + "/", zos, false);
+          } else {
+            filePath = relativePath + f.getName();
+            if (!filePath.equals("META-INF/MANIFEST.MF")) {
+              ZipEntry anEntry = new ZipEntry(filePath);
+              copyToZipStream(f, anEntry, zos);
+            }
+          }
+        }
+      }
+
+    }
+  }
+
+  private static void createJar(File dir, File jarFile) throws IOException {
+    Preconditions.checkNotNull(dir, "dir");
+    Preconditions.checkNotNull(jarFile, "jarFile");
+    File jarDir = jarFile.getParentFile();
+    if (!jarDir.exists() && !jarDir.mkdirs()) {
+      throw new IOException(MessageFormat.format("could not create dir [{0}]", jarDir));
+    } else {
+      FileOutputStream fos = new FileOutputStream(jarFile);
+      Throwable var4 = null;
+
+      try {
+        JarOutputStream jos = new JarOutputStream(fos);
+        Throwable var6 = null;
+
+        try {
+          jarDir(dir, "", jos);
+        } catch (Throwable var29) {
+          var6 = var29;
+          throw var29;
+        } finally {
+          if (jos != null) {
+            if (var6 != null) {
+              try {
+                jos.close();
+              } catch (Throwable var28) {
+                var6.addSuppressed(var28);
+              }
+            } else {
+              jos.close();
+            }
+          }
+
+        }
+      } catch (Throwable var31) {
+        var4 = var31;
+        throw var31;
+      } finally {
+        if (fos != null) {
+          if (var4 != null) {
+            try {
+              fos.close();
+            } catch (Throwable var27) {
+              var4.addSuppressed(var27);
+            }
+          } else {
+            fos.close();
+          }
+        }
+
+      }
+
+    }
+  }
+
+  public static String getJar(Class klass) {
+    Preconditions.checkNotNull(klass, "klass");
+    ClassLoader loader = klass.getClassLoader();
+    if (loader != null) {
+      String class_file = klass.getName().replaceAll("\\.", "/") + ".class";
+
+      try {
+        Enumeration itr = loader.getResources(class_file);
+
+        while(itr.hasMoreElements()) {
+          URL url = (URL)itr.nextElement();
+          String path = url.getPath();
+          if (path.startsWith("file:")) {
+            path = path.substring("file:".length());
+          }
+
+          path = URLDecoder.decode(path, "UTF-8");
+          if ("jar".equals(url.getProtocol())) {
+            path = URLDecoder.decode(path, "UTF-8");
+            return path.replaceAll("!.*$", "");
+          }
+
+          if ("file".equals(url.getProtocol())) {
+            String klassName = klass.getName();
+            klassName = klassName.replace(".", "/") + ".class";
+            path = path.substring(0, path.length() - klassName.length());
+            File baseDir = new File(path);
+            File testDir = new File(System.getProperty("test.build.dir", "target/test-dir"));
+            testDir = testDir.getAbsoluteFile();
+            if (!testDir.exists()) {
+              testDir.mkdirs();
+            }
+
+            File tempJar = File.createTempFile("hadoop-", "", testDir);
+            tempJar = new File(tempJar.getAbsolutePath() + ".jar");
+            tempJar.deleteOnExit();
+            createJar(baseDir, tempJar);
+            return tempJar.getAbsolutePath();
+          }
+        }
+      } catch (IOException var10) {
+        throw new RuntimeException(var10);
+      }
+    }
+
+    return null;
+  }
+}

--- a/emr-tablestore/src/main/java/com/aliyun/openservices/tablestore/hive/TableMapReduceUtils.java
+++ b/emr-tablestore/src/main/java/com/aliyun/openservices/tablestore/hive/TableMapReduceUtils.java
@@ -1,0 +1,152 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.aliyun.openservices.tablestore.hive;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.util.StringUtils;
+
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLDecoder;
+import java.util.*;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Borrowed from https://github.com/apache/hbase/blob/master/hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/TableMapReduceUtil.java
+ */
+public class TableMapReduceUtils {
+    private static final Log LOG = LogFactory.getLog(TableMapReduceUtils.class);
+
+  public static void addDependencyJars(Job job) throws IOException {
+    try {
+      addDependencyJarsForClasses(job.getConfiguration(), job.getMapOutputKeyClass(), job.getMapOutputValueClass(),
+        job.getInputFormatClass(), job.getOutputKeyClass(), job.getOutputValueClass(), job.getOutputFormatClass(),
+        job.getPartitionerClass(), job.getCombinerClass());
+    } catch (ClassNotFoundException var2) {
+      throw new IOException(var2);
+    }
+  }
+
+  public static void addDependencyJarsForClasses(Configuration conf, Class... classes) throws IOException {
+    FileSystem localFs = FileSystem.getLocal(conf);
+    Set<String> jars = new HashSet();
+    jars.addAll(conf.getStringCollection("tmpjars"));
+    Map<String, String> packagedClasses = new HashMap();
+    Class[] var5 = classes;
+    int var6 = classes.length;
+
+    for(int var7 = 0; var7 < var6; ++var7) {
+      Class<?> clazz = var5[var7];
+      if (clazz != null) {
+        Path path = findOrCreateJar(clazz, localFs, packagedClasses);
+        if (path == null) {
+          LOG.warn("Could not find jar for class " + clazz + " in order to ship it to the cluster.");
+        } else if (!localFs.exists(path)) {
+          LOG.warn("Could not validate jar file " + path + " for class " + clazz);
+        } else {
+          jars.add(path.toString());
+        }
+      }
+    }
+
+    if (!jars.isEmpty()) {
+      conf.set("tmpjars", StringUtils.arrayToString((String[])jars.toArray(new String[jars.size()])));
+    }
+  }
+
+  private static Path findOrCreateJar(Class<?> my_class, FileSystem fs, Map<String, String> packagedClasses) throws IOException {
+    String jar = findContainingJar(my_class, packagedClasses);
+    if (null == jar || jar.isEmpty()) {
+      jar = getJar(my_class);
+      updateMap(jar, packagedClasses);
+    }
+
+    if (null != jar && !jar.isEmpty()) {
+      LOG.debug(String.format("For class %s, using jar %s", my_class.getName(), jar));
+      return (new Path(jar)).makeQualified(fs);
+    } else {
+      return null;
+    }
+  }
+
+  private static String findContainingJar(Class<?> my_class, Map<String, String> packagedClasses) throws IOException {
+    ClassLoader loader = my_class.getClassLoader();
+    String class_file = my_class.getName().replaceAll("\\.", "/") + ".class";
+    if (loader != null) {
+      Enumeration itr = loader.getResources(class_file);
+
+      while(itr.hasMoreElements()) {
+        URL url = (URL)itr.nextElement();
+        if ("jar".equals(url.getProtocol())) {
+          String toReturn = url.getPath();
+          if (toReturn.startsWith("file:")) {
+            toReturn = toReturn.substring("file:".length());
+          }
+
+          toReturn = toReturn.replaceAll("\\+", "%2B");
+          toReturn = URLDecoder.decode(toReturn, "UTF-8");
+          return toReturn.replaceAll("!.*$", "");
+        }
+      }
+    }
+
+    return (String)packagedClasses.get(class_file);
+  }
+
+  private static String getJar(Class<?> my_class) {
+    String ret = null;
+
+    try {
+      ret = JarFinder.getJar(my_class);
+      return ret;
+    } catch (Exception var3) {
+      throw new RuntimeException("getJar invocation failed.", var3);
+    }
+  }
+
+  private static void updateMap(String jar, Map<String, String> packagedClasses) throws IOException {
+    if (null != jar && !jar.isEmpty()) {
+      ZipFile zip = null;
+
+      try {
+        zip = new ZipFile(jar);
+        Enumeration iter = zip.entries();
+
+        while(iter.hasMoreElements()) {
+          ZipEntry entry = (ZipEntry)iter.nextElement();
+          if (entry.getName().endsWith("class")) {
+            packagedClasses.put(entry.getName(), jar);
+          }
+        }
+      } finally {
+        if (null != zip) {
+          zip.close();
+        }
+
+      }
+
+    }
+  }
+}

--- a/emr-tablestore/src/main/java/com/aliyun/openservices/tablestore/hive/TableStoreStorageHandler.java
+++ b/emr-tablestore/src/main/java/com/aliyun/openservices/tablestore/hive/TableStoreStorageHandler.java
@@ -18,11 +18,15 @@
 
 package com.aliyun.openservices.tablestore.hive;
 
+import java.util.LinkedHashSet;
 import java.util.Properties;
 import java.util.Map;
+import java.util.Set;
 
 import com.aliyun.openservices.tablestore.hadoop.Credential;
 import com.aliyun.openservices.tablestore.hadoop.Endpoint;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.util.StringUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.apache.hadoop.hive.ql.metadata.DefaultStorageHandler;
@@ -81,11 +85,12 @@ public class TableStoreStorageHandler extends DefaultStorageHandler {
 
     @Override public void configureJobConf(TableDesc tableDesc,
         org.apache.hadoop.mapred.JobConf jobConf) {
-        Properties from = tableDesc.getProperties();
-        logger.debug("TableDesc: {}", from);
-        logger.debug("job conf: {}", jobConf);
-        jobConf.setJarByClass(TableStoreStorageHandler.class);
-        {
+        try {
+            Properties from = tableDesc.getProperties();
+            logger.debug("TableDesc: {}", from);
+            logger.debug("job conf: {}", jobConf);
+            jobConf.setJarByClass(TableStoreStorageHandler.class);
+
             String accessKeyId = from.getProperty(TableStoreConsts.ACCESS_KEY_ID);
             if (accessKeyId == null) {
                 logger.error("{} is required.", TableStoreConsts.ACCESS_KEY_ID);
@@ -102,8 +107,7 @@ public class TableStoreStorageHandler extends DefaultStorageHandler {
                 from.getProperty(TableStoreConsts.SECURITY_TOKEN));
             com.aliyun.openservices.tablestore.hadoop.TableStoreInputFormat
                 .setCredential(jobConf, cred);
-        }
-        {
+
             String endpoint = from.getProperty(TableStoreConsts.ENDPOINT);
             if (endpoint == null) {
                 logger.error("{} is required.", TableStoreConsts.ENDPOINT);
@@ -121,8 +125,7 @@ public class TableStoreStorageHandler extends DefaultStorageHandler {
             }
             com.aliyun.openservices.tablestore.hadoop.TableStoreInputFormat
                 .setEndpoint(jobConf, ep);
-        }
-        {
+
             String table = from.getProperty(TableStoreConsts.TABLE_NAME);
             if (table == null) {
                 logger.error("{} is required.", TableStoreConsts.TABLE_NAME);
@@ -131,8 +134,7 @@ public class TableStoreStorageHandler extends DefaultStorageHandler {
             }
             com.aliyun.openservices.tablestore.hadoop.TableStoreOutputFormat
                 .setOutputTable(jobConf, table);
-        }
-        {
+
             String t = from.getProperty(TableStoreConsts.MAX_UPDATE_BATCH_SIZE);
             if (t != null) {
                 try {
@@ -145,14 +147,21 @@ public class TableStoreStorageHandler extends DefaultStorageHandler {
                     }
                     com.aliyun.openservices.tablestore.hadoop.TableStoreOutputFormat
                         .setMaxBatchSize(jobConf, batchSize);
-                } catch(NumberFormatException ex) {
+                } catch (NumberFormatException ex) {
                     logger.error("{} must be a positive integer.",
                         TableStoreConsts.MAX_UPDATE_BATCH_SIZE);
                     throw new IllegalArgumentException(
                         TableStoreConsts.MAX_UPDATE_BATCH_SIZE + " must be a positive integer.");
                 }
             }
+
+            Set<String> merged = new LinkedHashSet<String>(jobConf.getStringCollection("tmpjars"));
+            Job copy = new Job(jobConf);
+            TableMapReduceUtils.addDependencyJars(copy);
+            merged.addAll(copy.getConfiguration().getStringCollection("tmpjars"));
+            jobConf.set("tmpjars", StringUtils.arrayToString(merged.toArray(new String[0])));
+        } catch (Exception e) {
+            throw new RuntimeException(e);
         }
     }
-
 }

